osu.Game/Beatmaps/BeatmapManager.cs

```diff
@@ -400,6 +411,82 @@ public void UndeleteAll()
 
         public Task Export(BeatmapSetInfo beatmap) => beatmapExporter.ExportAsync(beatmap.ToLive(Realm));
 
+        /// <summary>
+        /// Creates a copy of the <see cref="BeatmapSetInfo"/> and converts all beatmaps to legacy format, then exports it as a legacy package.
+        /// </summary>
+        /// <param name="beatmapSetInfo"></param>
+        /// <returns></returns>
+        public Task ExportLegacy(BeatmapSetInfo beatmapSetInfo)
+        {
+            // Create a clone of the original beatmap set which we will convert to legacy format and then export
+            var clone = new BeatmapSetInfo(beatmapSetInfo.Beatmaps.Select(b => b.Clone()));
+            clone.Files.AddRange(beatmapSetInfo.Files.Select(f => new RealmNamedFileUsage(f.File, f.Filename)));
+
+            // convert all beatmaps in the cloned beatmap set to legacy format
+            foreach (var beatmapInfo in clone.Beatmaps)
+            {
+                beatmapInfo.BeatmapSet = clone;
+                beatmapInfo.ID = Guid.NewGuid();
+
+                var file = beatmapInfo.File;
+                if (file == null)
+                    continue;
+
+                var beatmapContent = new LegacyBeatmapDecoder().Decode(new LineBufferedReader(RealmFileStore.Storage.GetStream(file.File.GetStoragePath())));
+                var beatmapSkin = new LegacySkin(new SkinInfo(), null!)
+                {
+                    Configuration = new LegacySkinDecoder().Decode(new LineBufferedReader(RealmFileStore.Storage.GetStream(file.File.GetStoragePath())))
+                };
+
+                Realm.Realm.Write(realm =>
+                {
+                    using var stream = new MemoryStream();
+                    convertAndEncodeLegacyBeatmap(beatmapContent, beatmapSkin, stream);
+
+                    beatmapInfo.MD5Hash = stream.ComputeMD5Hash();
+                    beatmapInfo.Hash = stream.ComputeSHA2Hash();
+
+                    file.File = RealmFileStore.Add(stream, realm).Detach();
+                });
+            }
+
+            return legacyBeatmapExporter.ExportAsync(new RealmLiveUnmanaged<BeatmapSetInfo>(clone));
+        }
+
+        private void convertAndEncodeLegacyBeatmap(IBeatmap beatmapContent, ISkin beatmapSkin, Stream stream)
+        {
+            // Convert beatmap elements to be compatible with legacy format
+            // So we truncate time and position values to integers, and convert paths with multiple segments to bezier curves
+            foreach (var controlPoint in beatmapContent.ControlPointInfo.AllControlPoints)
+                controlPoint.Time = Math.Floor(controlPoint.Time);
+
+            foreach (var hitObject in beatmapContent.HitObjects)
+            {
+                hitObject.StartTime = Math.Floor(hitObject.StartTime);
```

Comment:
It would be applicable for spinners and hold notes, but not for sliders because those calculate the duration via the path length and velocity. I don't know if I want to complicate this by telling it exactly which classes it should truncate.

Reply:
Judging from

https://github.com/ppy/osu/blob/66269be1894111086765f0f4c378cd87af802dac/osu.Game/Beatmaps/Formats/LegacyBeatmapEncoder.cs#L380-L391

it should be as simple as doing `hitObject is IHasDuration && hitObject is not IHasPath`, no?

Reply:
actually true